paasta_tools/kubernetes/application/controller_wrappers.py

```diff
@@ -188,14 +203,173 @@ def deep_delete(self, kube_client: KubeClient) -> None:
+            * 100
+        )
+        # TODO support bespoke
+        if (
+            self.get_soa_config().get("autoscaling", {}).get("decision_policy", "")
+            == "bespoke"
+        ):
+            return
+        elif metrics_provider == "mesos_cpu":
+            metrics.append(
+                V2beta1MetricSpec(
+                    type="Resource",
+                    resource=V2beta1ResourceMetricSource(
+                        name="cpu", target_average_utilization=target
+                    ),
+                )
+            )
+        elif metrics_provider == "http":
+            metrics.append(
+                V2beta1MetricSpec(
+                    type="Pods",
+                    pods=V2beta1PodsMetricSource(
+                        metric_name="http", target_average_value=target
+                    ),
+                )
+            )
+        elif metrics_provider == "uwsgi":
+            metrics.append(
+                V2beta1MetricSpec(
+                    type="Pods",
+                    pods=V2beta1PodsMetricSource(
+                        metric_name="uwsgi", target_average_value=target
+                    ),
+                )
+            )
+        else:
+            self.logging.error("Wrong metrics specified")
+            return
+
+        body = V2beta1HorizontalPodAutoscaler(
+            kind="HorizontalPodAutoscaler",
+            metadata=V1ObjectMeta(
+                name=self.item.metadata.name, namespace=self.item.metadata.namespace
+            ),
+            spec=V2beta1HorizontalPodAutoscalerSpec(
+                max_replicas=max_replicas,
+                min_replicas=min_replicas,
+                metrics=metrics,
+                scale_target_ref=V2beta1CrossVersionObjectReference(
+                    kind="Deployment", name=self.item.metadata.name
+                ),
+            ),
+        )
+        self.logging.debug(body)
+        if not hpa_exists:
+            self.logging.info(
+                "Creating new HPA for {metrics_provider} {self.item.metadata.name}/name in {self.item.metadata.namespace}"
+            )
+            kube_client.autoscaling.create_namespaced_horizontal_pod_autoscaler(
+                namespace=self.item.metadata.namespace, body=body, pretty=True
+            )
+        else:
+            self.logging.info(
+                "Updating new HPA for {metrics_provider} {self.item.metadata.name}/name in {self.item.metadata.namespace}/namespace"
+            )
+            kube_client.autoscaling.patch_namespaced_horizontal_pod_autoscaler(
+                name=self.item.metadata.name,
+                namespace=self.item.metadata.namespace,
+                body=body,
+                pretty=True,
+            )
+
+    def exists_hpa(self, kube_client: KubeClient) -> bool:
+        try:
+            kube_client.autoscaling.read_namespaced_horizontal_pod_autoscaler(
```

Comment:
nit
Checking existence of something by using `read_*` and comparing HTTP error with 404 is quite fragile.
We could get 404 not only because the object doesn't exist, but also because of an API change, because of incorrectly configured LB, or million other reasons.
It's much better to use `list_*` with proper set of selector and compare amount of results with `1`.